Drees 102 pmsi flattening

alecoules opened the pull request:
Hi,

Our pull request is ready for the flattening of MCO tables and SSR/RIP tables. The issue was that MCO, SSR and RIP can't be handled in the same way as DCIR. 
For the PMSI, we have to first join the two central tables (MCO_B, MCO_C) and then for each other table join to the result and stack every outputs (code is in **PMSIFlatTable.scala**). 
For SSR and RIP, it's the same thing but we have to be be careful because the keys aren't the same for the first join and the following (code is in **SSRRIPFlatTable.scala**).
We had to add new keys in the conf files : **joinKeysPatient**, **pmsiPatientTableName**. The technique chosen by the algorithm to join the input tables depends of the existence of these keys in the conf file (code is in **FlatteningConfig.scala** and in **Flattening.scala**).
We also wrote tests (**PMSIFlatTableSuite.scala** and **SSRRIPFlatTableSuite.scala**) and add csv, parquet and conf files for these tests.
We tested our code on the Drees sample (4M persons during 9 years).

Arnaud Lecoules

arnaudlecoules committed:
DREES-102-PMSIFlattening: Writing Tests for new flattening methods

Preparation for merge request

Error correction on src/test/scala/fr/polytechnique/cmap/cnam/flattening/FlatteningConfigSuite.scala

adding config files

adding PMSI testing

adding test_PMSI key in /Users/Arnaud/Desktop/SCALPEL-Flattening/src/main/resources/flattening/config/main.conf

renaming /Users/Arnaud/Desktop/SCALPEL-Flattening/src/main/scala/fr/polytechnique/cmap/cnam/flattening/SSRFlatTable.scala into /Users/Arnaud/Desktop/SCALPEL-Flattening/src/main/scala/fr/polytechnique/cmap/cnam/flattening/SSR-RPIFlatTable.scala

ajout de fichiers .csv SSR pour les tests

Rename /Users/Arnaud/Desktop/SCALPEL-Flattening/src/main/scala/fr/polytechnique/cmap/cnam/flattening/SSR-RPIFlatTable.scala

conf files for SSR test

RPI -> RIP

Writing test for SRRRIPFlatTable

adding missing parquet files for SSRRIPFlatTableSuite

fix typo in test files

sorted results for years

arnaudlecoules committed:
DREES-102-PMSIFlattening: Rewriting the code in order to create new classes inheritance

SSRRIPFlatTable now inherits from PMSIFlatTable

PMSIFlatTable now inherits from FlatTable

Modification from DREES-103-PrefixProblem

Correcting mergeSchemas in order to be more readable

adding refsToJoin to Flattening.scala to new flattening methods

ysebiat commented on src/main/scala/fr/polytechnique/cmap/cnam/flattening/PMSIFlatTable.scala:
```diff
@@ -0,0 +1,58 @@
+   * DataFrame, it created the column as empty
+   */
+  def mergeSchemas(myCols: Set[String], allCols: Set[String]) = {
+    allCols.toList.map(x => x match {
```
The variable x is ambiguous in this method. The name x captures the element you pass. Here is a suggestion:
```scala
  def mergeSchemas(myCols: Set[String], allCols: Set[String]): List[Column] = 
    allCols.toList.map(colName => if (myCols.contains(colName)) col(colName) else lit(null).alias(colName))
```

alecoules replied:
OK. I commited this change.

alecoules replied:
Subsidiary question : why did you prefere if/else versus match pattern ?

ysebiat replied:
I would prefer the match pattern for complex pattern matching. The example above is a boolean if around a test, which IMO is better suited for an if/else control.

ysebiat reviewed (changes requested):
A comment on the merge method.

arnaudlecoules committed:
DREES-102-PMSIFlattening: Writing new tests for new flattening methods (SSR/RIP) and in order to test new classes

adding a flattening test for MCO with PMSI logic

Error in FlatteningSuite where two test had the same name

adding SSR tables test FlatteningSuite

modifying tables for test and adding new ones for a forecoming test

Preparation of test tables

test for joinByYearAndDate for PMSI Logic

Still working on tests

adding files for the tests

minor change in FlatteningSuite.scala

Modifying SSRRIPFlatTableSuite.scala

Adding test for joinByYearAndDate in SSRRIPFlatTable

arnaudlecoules committed:
DREES-102-PMSIFlattening: Formal modifs for the PR

removing script FlatteningJoinOnly.scala still in test

Imports cleaning

removing unused conf from src/main/resources/flattening/config/main.conf

adding comment tosrc/main/scala/fr/polytechnique/cmap/cnam/flattening/Flattening.scala

formal modifs for the PR

minor modif

updating sbt version

ysebiat commented on src/main/resources/flattening/config/main.conf:
```diff
@@ -17,3 +17,39 @@ pio = {
 fraude = {
   include "fraude/env.conf"
 }
+
```
Do you think this configurations are necessary for the whole project?

alecoules replied:
I dropped every confs related to drees_echantillon

ysebiat commented on src/main/resources/flattening/config/ssr_month/env.conf:
```diff
@@ -0,0 +1,23 @@
+base_path = "target/test/output"
```
I think you can omit this file from the PR.

alecoules replied:
I'm using it for the tests

ysebiat commented on src/main/resources/flattening/config/ssr_month/ssr.conf:
```diff
@@ -0,0 +1,67 @@
+tables = [
```
I think you can omit this file from the PR.

alecoules replied:
I'm using it for the tests

ysebiat commented on src/main/scala/fr/polytechnique/cmap/cnam/flattening/Flattening.scala:
```diff
@@ -46,11 +46,33 @@ object Flattening {
       }.toList
   }
 
+  /**
```
Add documentation for the arguments.

alecoules replied:
Done

ysebiat commented on src/main/scala/fr/polytechnique/cmap/cnam/flattening/Flattening.scala:
```diff
@@ -46,11 +46,33 @@ object Flattening {
+          OperationMetadata(config.name, config.flatOutputPath.get, "flat_table", config.mainTableName :: config.tablesToJoin, config.joinKeys)
+        }
+        else {
+          logger.info("join SSR table " + config.name + " with Pmsi logic")
```
Better to say
```scala
logger.info("join table " + config.name + " with SSR-PMSI logic")
```

ysebiat commented on src/test/scala/fr/polytechnique/cmap/cnam/flattening/FlatteningSuite.scala:
```diff
@@ -5,6 +5,7 @@ package fr.polytechnique.cmap.cnam.flattening
 import org.apache.spark.sql._
 import fr.polytechnique.cmap.cnam.SharedContext
 import fr.polytechnique.cmap.cnam.utilities.DFUtils._
+import org.apache.spark.sql.functions.col
```
The order of the import must follow the following order:
```scala
java
javax
scala
all other imports
org.apache.spark
fr.polytechnique

ysebiat commented on src/test/scala/fr/polytechnique/cmap/cnam/flattening/PMSIFlatTableSuite.scala:
```diff
@@ -0,0 +1,68 @@
+
+class PMSIFlatTableSuite extends SharedContext {
+
+  "WriteAsParquet" should "flatten MCO and write it in the correct path" in {
```
```writeAsParquet```

ysebiat commented on src/test/scala/fr/polytechnique/cmap/cnam/flattening/PMSIFlatTableSuite.scala:
```diff
@@ -0,0 +1,68 @@
+    assert(expectedDF_sorted.count() == result_sorted.count())
+  }
+
+  "WriteAsParquetMonth" should "flatten MCO monthly and write it in the correct path" in {
```
```writeAsParquetMonth```

ysebiat reviewed (changes requested):
Some remarks about
1. Order of the imports.
2. Some configuration files that I think are not necessary for the PR.

ysebiat reviewed (approved):
Ready to merge.
